### PR TITLE
add metadata api 

### PR DESCRIPTION
**Pull request by saber-wang**
# Description

_Please explain the changes you've made_

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #883

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

File: src/Dapr.Client/DaprClient.cs [hunk 1]
```diff
@@ -330,7 +330,25 @@ public HttpRequestMessage CreateInvokeMethodRequest<TRequest>(string appId, stri
         /// <param name="cancellationToken">A <see cref="CancellationToken" /> that can be used to cancel the operation.</param>
         /// <returns>A <see cref="Task" /> that will return when the operation has completed.</returns>
         public abstract Task WaitForSidecarAsync(CancellationToken cancellationToken = default);
-        
+
+        /// <summary>
+        ///Perform service returns information about the sidecar allowing runtime discoverability. 
+        ///The metadata endpoint returns a list of the components loaded, 
+        ///the activated actors (if present) and attributes with information attached.
```
Comment (halspang) [inline src/Dapr.Client/DaprClient.cs:337]: There should be a space between `//` and the first letter.

Comment (halspang) [reply src/Dapr.Client/DaprClient.cs:337]: I'd also update the comment to:

```csharp
/// <summary>
/// Calls the sidecar's metadata endpoint which returns information including:
/// <list type="bullet">
/// <item>
/// <description>The sidecar's ID.</description>
/// </item>
/// <item>
/// <description>The registered/active actors if any.</description>
/// </item>
/// <item>
/// <description>Registered components including name, type, version, and information on capabilities if present.</description>
/// </item>
/// <item>
/// <description>Any extended metadata that has been set via <see cref="SetMetadataAsync"/></description>
/// </item>
/// </list>
/// </summary>
```

File: src/Dapr.Client/DaprClient.cs [hunk 2]
```diff
@@ -330,7 +330,25 @@ public HttpRequestMessage CreateInvokeMethodRequest<TRequest>(string appId, stri
...
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> that can be used to cancel the operation.</param>
+        /// <returns>A <see cref="Task{T}" /> that will return the value when the operation has completed.</returns>
+        public abstract Task<DaprMetadata> GetMetadataAsync(CancellationToken cancellationToken = default);
+
+        /// <summary>
+        ///Perform service Perform service adds a custom label to the Dapr sidecar information stored.
```
Comment (halspang) [inline src/Dapr.Client/DaprClient.cs:344]: Reduce this too:

```csharp
Add extended metadata to the Dapr sidecar.
```

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 3]
```diff
@@ -333,6 +333,45 @@ public override async Task WaitForSidecarAsync(CancellationToken cancellationTok
...
+            try
+            {
+                var response = await this.httpClient.SendAsync(request, cancellationToken);
+                response.EnsureSuccessStatusCode();
+            }
+            catch (HttpRequestException ex)
+            {
+                throw new DaprException("Set metadata operation failed: the Dapr endpoint indicated a failure. See InnerException for details.", ex);
+            }
+        }
```
Comment (halspang) [inline src/Dapr.Client/DaprClientGrpc.cs:373]: Both of these should use the internal client. The metadata API is specified in the proto, so we should use it's generated client.

File: src/Dapr.Client/DaprMetadata.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        /// <param name="extended">The list of custom attributes as key-value pairs, where key is the attribute name.</param>
+        /// <param name="components">The loaded  components metadata.</param>
+        public DaprMetadata(string id, IReadOnlyList<DaprActorMetadata> actors, IReadOnlyDictionary<string, string> extended, IReadOnlyList<DaprComponentsMetadata> components)
+        {
+            Id = id;
+            Actors = actors;
+            Extended = extended;
+            Components = components;
+        }
+        /// <summary>
```
Comment (halspang) [inline src/Dapr.Client/DaprMetadata.cs:37]: Nit: Needs a new line.

File: src/Dapr.Client/DaprMetadata.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        /// Gets the list of custom attributes as key-value pairs, where key is the attribute name.
+        /// </summary>
+        public IReadOnlyDictionary<string, string> Extended { get; }
+
+        /// <summary>
+        ///  Gets the loaded  components metadata.
+        /// </summary>
+        public IReadOnlyList<DaprComponentsMetadata> Components { get; }
+
```
Comment (halspang) [inline src/Dapr.Client/DaprMetadata.cs:56]: Nit: Does not need a new line.

File: src/Dapr.Client/DaprMetadata.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        /// <param name="version">The component version.</param>
+        /// <param name="capabilities">The supported capabilities for this component type and version.</param>
+        public DaprComponentsMetadata(string name, string type, string version, string[] capabilities)
+        {
+            Name = name;
+            Type = type;
+            Version = version;
+            Capabilities = capabilities;
+        }
+
```
Comment (halspang) [inline src/Dapr.Client/DaprMetadata.cs:105]: Nit: Extra line.

**Review by halspang — CHANGES_REQUESTED** [review]

File: src/Dapr.Client/DaprClientGrpc.cs (hunk 3 above)
Comment (saber-wang) [reply src/Dapr.Client/DaprClientGrpc.cs:373]: The grpc API is currently unavailable If grpc API is used, it is not available in the lower version of daprd, see https://github.com/dapr/dapr/issues/4941

Comment (saber-wang) [reply src/Dapr.Client/DaprClientGrpc.cs:373]: Does it need to be modified here? If need to modify, I will revert to the first [commit](https://github.com/dapr/dotnet-sdk/pull/917/commits/48fb236c8649f02c1ae1c637d3f9aea4be45fd7a)

**Pull request closed without merging**